extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/MSQTasks.java

```diff
@@ -65,9 +71,26 @@ public static String controllerTaskId(@Nullable final String queryId)
   /**
    * Returns a worker task ID given a SQL query id.
    */
-  public static String workerTaskId(final String controllerTaskId, final int workerNumber)
+  public static String workerTaskId(final String controllerTaskId, final int workerNumber, final int retryCount)
   {
-    return StringUtils.format("%s-worker%d", controllerTaskId, workerNumber);
+    return StringUtils.format("%s-worker%d_%d", controllerTaskId, workerNumber, retryCount);
+  }
+
+  /**
+   * Extract worker from taskId or throw exception if unable to parse out the worker.
+   */
+  public static int workerFromTaskId(final String taskId)
+  {
+    final Matcher matcher = WORKER_PATTERN.matcher(taskId);
+    if (matcher.matches()) {
+      return Integer.parseInt(matcher.group(1));
```

Comment:
If we can use the named group in the regex, then we won't have to use the stray reference to `1` and the code would remain valid irrespective of regex.
```suggestion
      return Integer.parseInt(matcher.group("workerNumber"));
```